File: person.py
import math
import numpy as np

class Person:
    def __init__(self,i, posx, posy, objx, objy, v, t_recovery, t_incubation, quarantined):
        # movement speed
        self.v = v
        # target position
        self.objx = objx
        self.objy = objy
        #ID and name
        self.indice = i
        self.name   = "Person "+str(i)
        #State: Susceptible, Infected or Retired
        self.infectious = False
        self.exposed    = False
        self.suceptible = True
        self.removed    = False
        #Current position
        self.posx = posx
        self.posy = posy
        #is it fixed (in quarantine)?
        self.quaritined = quarantined

        # displacement per iteration
        if self.quaritined:
            self.deltax = 0
            self.deltay = 0
        else:
            self.deltax = (self.objx - self.posx) / self.v
            self.deltay = (self.objy - self.posy) / self.v
        #time in which the person was infected
        self.t_contaminated = -1
        #time that the infection lasts, recover time
        self.t_recovery = t_recovery
        #time before the person becomes infection, incubation period
        self.t_incubation = t_incubation


    def __str__(self):
        return self.name+" en posicin "+str(self.posx)+", "+str(self.posy)

    # The person becomes infected but is not yet infectious
    def incubate(self,i):
        self.infectious = False
        self.exposed    = True
        self.suceptible = False
        self.removed    = False
        self.t_contaminated = i

    # The person becomes infectious after the incubation period has passed
    def infect(self):
        self.infectious = True
        self.exposed    = False
        self.suceptible = False
        self.removed    = False

    def remove(self):
        #heal
        self.removed=True
        self.suceptible=False
        self.infectious=False

    def set_objective(self,objx,objy):
        #this function is used to create a new target position
        self.objx=objx
        self.objy=objy
        if self.quaritined:
            self.deltax = 0
            self.deltay=0
        else:
            self.deltax = (self.objx - self.posx) / self.v
            self.deltay = (self.objy - self.posy) / self.v
        print("New objective  ", self.objx,self.objy,"  ",self.indice)

    def check_contamination(self,i):
        #this function is used to heal the person if the established infection time has passed
        if self.t_contaminated>-1:

            if i-self.t_contaminated-self.t_incubation >self.t_recovery:
                self.remove()

            elif i-self.t_contaminated > self.t_incubation:
                self.infect()


    def update_pos(self, n_posx, n_posy):
        #this funcion animates the movement
        if(n_posx==0 and n_posy==0):
            self.posx=self.posx+self.deltax
            self.posy=self.posy+self.deltay
        else:
            self.posx=n_posx
            self.posy=n_posy

        if abs(self.posx-self.objx)<3 and abs(self.posy-self.objy)<3:
            self.set_objective(np.random.random()*100, np.random.random()*100)
        if self.posx>100:
            self.posx=100
        if self.posy>100:
            self.posy=100
        if self.posx<0:
            self.posx=0
        if self.posy<0:
            self.posy=0

    def get_color(self):
        if self.infectious:
            return 'red'
        if self.exposed:
            return 'orange'
        if self.suceptible:
            return 'blue'
        if self.removed:
            return 'gray'

    def get_pos(self):
        return (self.posx,self.posy)

    def get_dist(self,x,y):
        #this funcion calculates the distance between this person an another.
        return math.sqrt((self.posx-x)**2+(self.posy-y)**2)

File: person.c
#include"person.h"


void incubate(Person *person, int i)
{
    person->infectious = false;
    person->exposed = true;
    person->susceptible= false;
    person->removed = false;
    person->t_contaminated = i;
}

void infect(Person *person)
{
    person->infectious = true;
    person->exposed = false;
    person->susceptible = false;
    person->removed = false;
}

void remove_p(Person *person, int i)
{
    person->t_removed = i;
    person->t_contaminated = -1;
    person->removed = true;
    person->susceptible = false;
    person->infectious = false;
}

void set_objective(Person *person, const int obj_x, const int obj_y)
{
    person->obj_x = obj_x;
    person->obj_y = obj_y;
    if(person->quarantined)
    {
        person->delta_x = 0;
        person->delta_y = 0;
    }
    else
    {
        person->delta_x = ( person->obj_x - person->pos_x ) / person->speed;
        person->delta_y = ( person->obj_y - person->pos_y ) / person->speed;
    }
}

void check_contamination(Person *person, int i)
{
    if(person->t_contaminated > -1)
    {
        if ((i - person->t_contaminated - person->t_incubation) > person->t_recovery )
        {
            remove_p(person, i);
        }
        else if ((i - person->t_contaminated ) > person->t_incubation)
        {
            infect(person);
        }
        

    }
}

void update_pos(Person *person, double n_pos_x, double n_pos_y)
{
    if(n_pos_x == 0 && n_pos_y == 0)
    {
        person->pos_x = person->pos_x + person->delta_x;
        person->pos_y = person->pos_y + person->delta_y;
    }
    else
    {
        person->pos_x = n_pos_x;
        person->pos_y = n_pos_y;
    }

    if(abs(person->pos_x - person->obj_x) < 3 && abs(person->pos_y - person->obj_y) < 3 )
    {
        set_objective(person, random_bounded_num(1,100), random_bounded_num(1,100));
    }
    if (person->pos_x > 100) person->pos_x = 100;
    if (person->pos_y > 100) person->pos_y = 100;
    if (person->pos_x < 0) person->pos_x = 0;
    if (person->pos_y < 0) person->pos_y = 0;
}

double get_distance(Person *p1, Person *p2)
{
    return sqrt(pow(p1->pos_x - p2->pos_x, 2) + pow(p1->pos_y - p2->pos_y, 2));
}

int random_bounded_num(int lower, int upper)
{
    return (rand() % (upper - lower + 1)) + lower; 
}
